### ISSUE TITLE
checkpoint commands group

### ISSUE THREAD

**Issue by buhtignew** [slimcoin-project/pacli#6]
I was trying to understand how `checkpoint set` command works and have run `checkpoint set -h` which has produced the following output:
```
Storing hash of block as a checkpoint to control re-orgs.
 Height: True Hash: {'code': -1, 'message': 'value is type bool, expected int'}
Stored checkpoint:
Label: True
Value: {'code': -1, 'message': 'value is type bool, expected int'}
```
Later on I've discovered that the `-h` flag in this context doesn't mean help. Maybe we should suppress the `-h` flag is this case requiring to select explicitly the `--height` and the `--help` flag or to align the `-h` with the other ones in our code, thus switching it to `--help` and asking user to be explicit is he wants to use the `--height` flag?

Now my output for `checkpoint show` is the following: 
```
  File "~/.local/bin/pacli", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "~/.local/lib/python3.11/site-packages/pacli/__main__.py", line 441, in main
    fire.Fire({
  File "~/.local/lib/python3.11/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/fire/core.py", line 475, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
                                ^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/pacli/extended_checkpoints.py", line 55, in show
    return retrieve_checkpoint(height=height)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/pacli/extended_checkpoints.py", line 90, in retrieve_checkpoint
    bheights = sorted([ int(h) for h in config["checkpoint"] ])
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/pacli/extended_checkpoints.py", line 90, in <listcomp>
    bheights = sorted([ int(h) for h in config["checkpoint"] ])
                        ^^^^^^
ValueError: invalid literal for int() with base 10: 'true'
```
The output of `checkpoint list` is:
```
["169505", "00000003a918e11d0c8cfc0304a8732a34935693f6867ec6e0792bbb3d71b4e9"]
["177497", "aea2aa2fd5bca2fb7167fa5bd736d9d6e6879fba803cfd48faa5b3c1a3f88710"]
["177646", "46ffaf25e8ea7f6496517f6e6fbe1e6b4db7a462c8b160eeac8baad672b65f19"]
["177654", "31f094fd3f6f361eba6dae2d064acdb6f075cf4029b3a3cfc0f18b78243fb522"]
["203590", "6b74ad26769e2a566a8cb3e031aa0f6aa69b2417a345240050f4bdac6f910fc2"]
["203607", "19da7a4b3b71068bad4fa153695813a71ee06594576cf430c6f22c5695b8d16f"]
["244222", "731db387e4c0ed95c94e62d80a2b92c5dde40d6e39a99315499b9f20a2510667"]
["true", {"code": -1, "message": "value is type bool, expected int"}]
```
Maybe we should add a check for the `checkpoint set` command in order it to avoid accepting strings instead of numbers as the checkpoint height.

I've run `checkpoint set true --delete --now` and was able to successfully eliminate the "true" checkpoint. The `checkpoint list` output is as following now:
```
["169505", "00000003a918e11d0c8cfc0304a8732a34935693f6867ec6e0792bbb3d71b4e9"]
["177497", "aea2aa2fd5bca2fb7167fa5bd736d9d6e6879fba803cfd48faa5b3c1a3f88710"]
["177646", "46ffaf25e8ea7f6496517f6e6fbe1e6b4db7a462c8b160eeac8baad672b65f19"]
["177654", "31f094fd3f6f361eba6dae2d064acdb6f075cf4029b3a3cfc0f18b78243fb522"]
["203590", "6b74ad26769e2a566a8cb3e031aa0f6aa69b2417a345240050f4bdac6f910fc2"]
["203607", "19da7a4b3b71068bad4fa153695813a71ee06594576cf430c6f22c5695b8d16f"]
["244222", "731db387e4c0ed95c94e62d80a2b92c5dde40d6e39a99315499b9f20a2510667"]
```
The `checkpoint show` doesn't report errors as well.

**Comment by d5000** [comment]
I've changed the `--height` flag now to` --blockheight` so it doesn't block the short form` -h` for `--help`.

I also removed the `--depth` parameter, the depth has to be entered simply after `--prune` / `-p`. So now we have a short form for all flags and options in this command.

I have to admit that until you started testing in this round I wasn't even aware of the short forms of the flags (-h instead of --help etc.) :D . So I may go through all commands and see if I can remove these "duplicates" at least for the more popular flags. Of course --help should always be able to be shortened to -h for example.

You have already solved the error message you got, congrats :) As you suggested, I have added a check that checkpoints can only be saved as a number, not other values. I'll have add similar checks to other commands as well.

(BTW: thank you for opening a new issue. In general I think each command should have an own issue, so we can close it rapidly if all problems are fixed. It can of course be re-opened at any time if another issue occurs. In the case of checkpoints, as they are a small group, it's ok to group all of them into one issue.)

**Comment by buhtignew** [comment]
> I have to admit that until you started testing in this round I wasn't even aware of the short forms of the flags (-h instead of --help etc.) :D . So I may go through all commands and see if I can remove these "duplicates" at least for the more popular flags. Of course --help should always be able to be shortened to -h for example.

There is another thing I've discovered, that may be useful for us to keep in mind, if for instance someone because of distraction uses `-layout` flag instead of `--layout` (or `-help` instead of `--help` or whatsoever) the flag will be taken into consideration and command executed with the flag  corresponding to the first letter of the word, `-l` or `-h` for instance. It's another reason why we shouldn't use abbreviations that are different from the initial letter of the word.
I'm wondering whether there is the possibility to contract the abbreviated flags into one string, like `--wallet --quiet -- send` into `-wqs` if so we probably should check whether the other letters of our full word flags are not being considered in case the user makes a mistake of omitting one minus sign (`-`) before the flag word and at least how catastrophic can be such a mistake in each specific case. But maybe the `-layout` word for instance is not considered as contracted short flags string because it contains the letters that are not corresponding to any shorted flag.

**Comment by d5000** [comment]
The blockheight variable is normally not needed to be used as a flag, i.e. you can simply use `checkpoint set HEIGHT` instead of `checkpoint set -b HEIGHT.`  Thus I've not mentioned the flag in the docstring. However the Fire help function shows it anyway. I know this is not ideal but I currently don't know if this can be suppressed.

The depth variable has to be given directly after -p / --prune. The pruning mode of `checkpoint set` does not use the blockheight variable at all, so it will be simply ignored. This is why the message and behaviour of the command didn't change when you changed the blockheight variable (the one immediately after `checkpoint set`).

You are however correct that there is something missing in the documentation (your bullet point 2): There is always a minimum of 5 checkpoints which will be kept, regardless of the depth, which are the 5 closest to the current block height. I'm sorry if this may have caused misunderstandings, don't know why I've not mentioned that in the docstring ... I have added it to the docstring, in the next version it will be pushed to the repo.

> I'm testing the checkpoint show command group and am wondering where the flag is being used at all. Do we really need it?

You have edited my own post here, I think by mistake :)  (Didn't know that was possible, but we should try to avoid that, I had almost overlooked that question)

I think my first sentence in this post answers your question - it's simply a variable, I haven't found a way where blockheight can't be mentioned in the flags section of the help (the only way would be to make it mandatory, but the prune option doesn't need it)

**Comment by buhtignew** [comment]
> The blockheight variable is normally not needed to be used as a flag, i.e. you can simply use checkpoint set HEIGHT instead of checkpoint set -b HEIGHT. Thus I've not mentioned the flag in the docstring. However the Fire help function shows it anyway. I know this is not ideal but I currently don't know if this can be suppressed.

Maybe we can mention is it the command description as optional [-b]?

> I could however change that to 5000 or even 10000 to be "safer", what do you mean?

I think 2000 is enough.

> You have edited my own post here, I think by mistake :) (Didn't know that was possible, but we should try to avoid that, I had almost overlooked that question)

I have indeed first edited your post then I was not finding the commend in my post and wrote it again.
I think it's a github's bug. If anyone can change the text of the other's comments he can put into the mouth of the other anything he's never said, that can be an issue in some context. Maybe it's because we both have extended privileges in slimcoin-project?

**Comment by d5000** [comment]
> Maybe we can mention it in the command description as optional [-b]?
I'd prefer clarifying it in the docstring in another way, because there's absolutely no reason why an user should input `-b`.

> I think it would be more comfortable and intuitive if instead of the [DEPTH] parameter the user were to input the [BLOCKHEIGHT] parameter for the checkpoint set --prune command. 

I disagree here, the depth allows you to use the pruning feature without having to care about the current block height. For example, if you think currently the blockchain is unstable, then you increase the depth to 5000 or 10000, and thus increasing re-org safety, without having to lookup the current block height first with `getblockcount`.

I can however provide the blockheight as an alternative way. This would work this way:

`pacli checkpoint set [-b] 100000 -p` -> would prune all checkpoints from 100000 onwards
`pacli checkpoint set -p 100000 `-> would use the depth, pruning the checkpoints until 100000 blocks before the current block.

This would however not make possible your suggestion to put the depth *before* the  `-p` option. I think however it's better that way, because so it becomes clear that the number behind the `-p` is something different (the depth).

I just realized that this option would be quite nice to have if for example a reorg occurs at block X, so people can fastly delete all checkpoints after X. So I think I'll add it.

> By the way what if we use the [DEPTH] parameter instead of [BLOCKHEIGHT] parameter to set and delete the checkpoints as well?

I disagree here too. If the blockchain just advances one block while the user types the command, then he would delete the wrong checkpoint or get an error. Instead the block height does not change and it's always clear which checkpoint is meant.

I don't think also the `checkpoint set` and `checkpoint show` commands (excluding the pruning option) would be used by many users. As with each transaction you create a new checkpoint is created, normal users would only need to know the prune command. The other variants are more for power users, for example if a reorg occurred and we need to give recommendations to the users via discord/forum.

> Using the checkpoint set [HEIGHT] -d -n command I've just been able to circumvent this rule, in fact here is the list of my checkpoints right now:

Correct, and that's intended. You can always delete checkpoints manually, so for example if you're on an orphan block and just store it as a checkpoint, you can delete it without hassle even if you currently have less than 5 checkpoints in the list. The "5-checkpoint rule" is only meant for the prune variant, as it potentially deletes many checkpoints.

> I think it would make sense to change the output message as well,

That's a good idea, while it needs a check of a couple of lines of code I think it may be useful.

> Maybe we should also introduce the rule of the minimum distance between the checkpoints for the checkpoint set command if not someone can create a lot of very close checkpoint defeating in that way the purpose of creating checkpoints.

Here I don't follow you, as even one block can make the difference if a reorg occurs. For example, you have checkpoint on block 50000 and 50001, and block 50001 gets orphaned, then you still have the checkpoint on 50000. The only thing the user loses if he stores too many checkpoints is HDD storage but very few bytes.

**Comment by buhtignew** [comment]
> I think we should simply refrain from it. 

Of course :-)

>>    Maybe we should also introduce the rule of the minimum distance between the checkpoints for the checkpoint set command if not someone can create a lot of very close checkpoint defeating in that way the purpose of creating checkpoints.

> Here I don't follow you, as even one block can make the difference if a reorg occurs. For example, you have checkpoint on block 50000 and 50001, and block 50001 gets orphaned, then you still have the checkpoint on 50000.

What if, for instance. someone creates 5 checkpoints one each next block while all the others checkpoints he has are all more than 2000 blocks old he can then activate the prune all command and delete all the old checkpoints. If all his new checkpoints are on the fork chain he basically wouldn't have valid checkpoints, would he? 

> I don't think also the checkpoint set and checkpoint show commands (excluding the pruning option) would be used by many users.

Maybe you are right here. My experience as tester is such that I look into the list of my checkpoints were only blockheights reported, then I try to prune them to see whether the prune flag works and to do so instead of copying the blockheight data from the `checkpoint list` command I need to make calculations to understand the depth I'm looking for, which is not that comfortable, while the normal user won't probably need to do so.

**Comment by d5000** [comment]
> What if, for instance. someone creates 5 checkpoints one each next block while all the others checkpoints he has are all more than 2000 blocks old he can then activate the prune all command and delete all the old checkpoints. If all his new checkpoints are on the fork chain he basically wouldn't have valid checkpoints, would he?

This is true, now I understand what you mean. I don't think this scenario will happen many times, but it is possible.

The question is if this would be really a problem. The reorg check only checks the most recent checkpoint. This means it checks if your own client performed a reorg (i.e. that it had a blockchain tip with the last blocks, but then exchanged it for another tip because the original tip was on a minority chain).

In the case you mention (Pacli having stored 5 very close checkpoints near the current blockchain tip), your client may be on the wrong chain but has still not reorganized. This case can't be detected by the reorg check because it would need information from other nodes. Slimcoin does this on its own, it decides whether it should reorganize or not. And once it reorganizes, then the reorg check would detect the reorg.

Old checkpoints are only useful in the extremely unlikely case that an long reorg occurs. You could then know how long the reorganized part of the chain is. I haven't implemented a check for this in pacli still. It could be an interesting addition for the `reorg_check` command, because it would help, in the case of a real attack or critical network fragmentation, to assess how big the damage is.

What can be done to make this always possible, is to always retain one single checkpoint outside of the chosen depth when pruning. For example, if you're on block 200000, have the default depth of 2000, and you have checkpoints on 197000, 199600, 199800, 199900, 199991 and 199995 then the 197000 checkpoint will stay there even if it's 3000 blocks away.

I've already implemented the alternative mode which allows to prune from/to a blockheight. It will be available in the next update after I test it (probably should be there until tomorrow Sunday).

**Comment by buhtignew** [comment]
> UPDATE: Pushed the commit with the enhancements to Github.

There is a typo in `checkpont set` docstring:
In the sentence "If BLOCKHEIGHT is given without -a/--above_block, checpoints until this block height are pruned.", in the word checkpoints a `k` is missing.

**Comment by d5000** [comment]
Thanks for spotting the typo, corrected :)

> You haven't added the explanation yet, have you?

No, I thought about it and thought it would confuse the user more than it would help. The problem is that many commands have this "problem", and thus I would have to add this "clarification" in all docstrings ...
If I don't find a way to eliminate the "FLAGS" section entirely (which would be my preferred option), then I think I add this clarification later to the manual.

About the "above block" option:

> My expectation was that the 341983 or the 342007 checkpoint would be deleted.

I think I have to rework this mode of the command, because the idea is to use it to delete checkpoints on a chain of orphan blocks. So if it only deletes one checkpoint it doesn't fulfill the user's expectations.
The reason why it only deleted the most recent checkpoint is the 5 minimum rule. It would make sense, I think, to remove this rule for the -a mode, so all checkpoints of orphan blocks get deleted.
This would of course have the "risk" that no checkpoints at all are preserved. So if this happens, perhaps it would make sense for the command to add at least one new checkpoint after having used this option.

Another idea is to add a `--force` option to override the 5-checkpoint minimum rule.

Which solution would you prefer?

> By the way, we don't have the flag --all for this group of commands, but still would it make sense to look for another beginning letter for the --above_block flag? For instance --greater_than or --on_top_of or --upward.

I'm open for suggestions here but these don't convince me (I don't have heard the adjective "great" for block heights, for example)

**Comment by buhtignew** [comment]
>>    You haven't added the explanation yet, have you?

> No, I thought about it and thought it would confuse the user more than it would help. The problem is that many commands have this "problem", and thus I would have to add this "clarification" in all docstrings ...
If I don't find a way to eliminate the "FLAGS" section entirely (which would be my preferred option), then I think I add this clarification later to the manual.

But what prevents you from putting the `-b` flag as optional in the command's description?

**Comment by d5000** [comment]
> I think --orphans_above [BLOCKHEIGHT] would make a great sense, because in fact it there is no need usually to prune checkpoints in the middle of the list unless there are orphans to deal with.

Great, I'll implement it that way.

The reason why I'm hesitant to include flags/options like `-b` which are not needed in the docstrings is that there is absolutely no need to know them for the user, and there are lots of them (at least half of all commands have this "problem"). Again, I would in reality like to remove the FLAGS section of the help, which is automatically generated by Fire, because it can lead to confusion. I have still not found this in the Fire documentation.

**Comment by buhtignew** [comment]
I don't know how can I test the --remove_orphans flag. The other commands and flags of the group seem to work, so I'm closing this issue.

**Comment by buhtignew** [comment]
I don't know how can I test the `--remove_orphans` flag at the moment.
_
I've tested `pacli checkpoint set [BLOCKHEIGHT] -p/--prune [DEPTH]` command with both BLOCKHEIGHT and DEPTH. Seems like DEPTH is ignored in such a case. Should we mention this behavior in help output?

**Comment by d5000** [comment]
> Should we mention this behavior in help output?
Done.
> I don't know how can I test the --remove_orphans flag at the moment.

I tested it in the following way:

Use `config set BLOCKHEIGHT ARBITRARY_STRING checkpoint -e` to generate an invalid checkpoint. It will be treated like an orphan (there is currently no check that the block hash is long enough and is really hexadecimal, I think this is unnecessary for now).

Only take care that BLOCKHEIGHT must be above the last current checkpoint mentioned in `checkpoint list`.

**Comment by buhtignew** [comment]
> Use config set BLOCKHEIGHT ARBITRARY_STRING checkpoint -e to generate an invalid checkpoint.

I was trying to test this command (in order to test ` --remove_orphans` later on) but was not lucky.
By running the `config set 353099 ARBITRARY_STRING checkpoint -e` command I'm getting:
```
Error: You have to provide a category if modifying the extended config file.
```
By running `checkpoint set  353099 ARBITRARY_STRING` or `checkpoint set  353099 ARBITRARY_STRING checkpoint -e` I'm getting the following error:
```
Traceback (most recent call last):
  File "~/.local/lib/python3.11/site-packages/pacli/config_extended.py", line 131, in delete_item
    print("WARNING: deleting item from category {}, label: {}, complete key: {}, value: {}".format(category, label, key, config[category][key]))
                                                                                                                         ~~~~~~~~~~~~~~~~^^^^^
KeyError: '353099'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "~/.local/bin/pacli", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "~/.local/lib/python3.11/site-packages/pacli/__main__.py", line 441, in main
    fire.Fire({
  File "~/.local/lib/python3.11/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/fire/core.py", line 475, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
                                ^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/pacli/extended_checkpoints.py", line 45, in set
    return ce.delete_item("checkpoint", str(blockheight), now=now, quiet=quiet)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/pacli/config_extended.py", line 134, in delete_item
    raise ei.PacliInputDataError("No item with this key. Nothing was deleted.")
pacli.extended_interface.PacliInputDataError: No item with this key. Nothing was deleted.
```

**Comment by d5000** [comment]
The order in the config set command has changed. It is now:

`config set BLOCKHEIGHT ARBITRARY_STRING -e checkpoint`

due to what was talked about in issue #20 . That was a bit unfortunate, I know ... sorry. 

> `checkpoint set  353099 ARBITRARY_STRING checkpoint -e`

The config commands have a completely different syntax then the checkpoint commands group. Mixing commands like this will almost never work.

> checkpoint set  353099 ARBITRARY_STRING
> KeyError: '353099'

You can't use checkpoint set this way - it will ignore the string you used, it only accepts the block number. That's why I mentioned `config set`.
Nevertheless, you discovered a bug here, which is due to a misunderstanding of mine how Fire interprets boolean values. I've fixed it in commit 1744b71.

**Comment by buhtignew** [comment]
> That was a bit unfortunate, I know ... sorry

I should've thought about it :-)

> The config commands have a completely different syntax then the checkpoint commands group. Mixing commands like this will almost never work.

I like `config set` command it's very powerful, I was not aware it would be possible to use it in this case.

Now I've tested `--remove_orphans` flag. Everything works fine. 

I think we can close this issue.
